### PR TITLE
Custom Scodec Codecs For Primitive Types

### PR DESCRIPTION
**Pull request by b-vennes**
## Purpose

- Implement binary serialization as functions that can be composed together to define serialization for complex types
- Perform serialization with less reliance on mutable objects

## Approach

- Implemented the Scodec `Encoder`, `Decoder`, and `Codec` algebras on the following types: `byte`, `byte-array`, `int`, `uint`, `short`, `ushort`, `intString`, `smallString`, `long`, `ulong`, `boolean`, and `option`

## Tests

- backwards compatibility tests between the new codecs and the existing `VLQByteStringReader` and `VLQByteStringWriter` classes
- encoding/decoding compatibility tests

## Closes

- #1545
- #1546
- #1564

### CODE REVIEW

File: common/src/main/scala/co/topl/utils/IdiomaticScalaTransition.scala [hunk 1]
```diff
@@ -58,5 +60,22 @@ private[topl] object IdiomaticScalaTransition {
     }
   }
 
-  object implicits extends ToValidatedOps with ToEitherOps with ToTryOps
+  trait AttemptOps[R] {
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/IdiomaticScalaTransition.scala:63]: I think there is a typo in this file for the toTryOps above this line. The implicit def is called toEitherOps and I think it should be toTryOps in keeping with all the others here.

Comment (scasplte2) [reply common/src/main/scala/co/topl/utils/IdiomaticScalaTransition.scala:63]: should we also add a toOptionOps? Or are we just converting all options to Either with a specified leftMap?

File: common/src/main/scala/co/topl/utils/codecs/binary/BooleanCodec.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+object BooleanCodec {
+
+  private val trueByte: Byte = 0x01
+  private val trueByteBitVector: BitVector = BitVector(trueByte)
+
+  private val falseByte: Byte = 0x00
+  private val falseByteBitVector: BitVector = BitVector(falseByte)
+
+  def decode(from: BitVector): Attempt[DecodeResult[Boolean]] =
+    if (from.length < byteSize) Attempt.failure(Err.insufficientBits(byteSize, from.length))
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/codecs/binary/BooleanCodec.scala:15]: a little confusing what scope `byteSize` is being pulled from

File: common/src/main/scala/co/topl/utils/codecs/binary/BooleanCodec.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    else {
+      val (headByteBits, remainder) = from.splitAt(byteSize)
+      Attempt.successful(DecodeResult(headByteBits === trueByteBitVector, remainder))
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/codecs/binary/BooleanCodec.scala:18]: I'd like this operator `===` addressed when we go over this

File: common/src/main/scala/co/topl/utils/codecs/binary/BooleanCodec.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+  def encode(value: Boolean): Attempt[BitVector] =
+    if (value) Attempt.successful(trueByteBitVector)
+    else Attempt.successful(falseByteBitVector)
+
+  val codec: Codec[Boolean] = new Codec[Boolean] {
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/codecs/binary/BooleanCodec.scala:25]: I'm sure we will but also please focus on this Codec[_] higher-kinded type

File: common/src/main/scala/co/topl/utils/codecs/binary/BooleanCodec.scala (hunk 2 above)
Comment (scasplte2) [reply common/src/main/scala/co/topl/utils/codecs/binary/BooleanCodec.scala:15]: realized that byteSize is just 8, not much we can do to make that more clear other than a comment in every file which is unfeasible.

File: common/src/main/scala/co/topl/utils/codecs/binary/BytesCodec.scala [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package co.topl.utils.codecs.binary
+
+import scodec.{Attempt, Codec, DecodeResult, Err, SizeBound}
+import scodec.bits.BitVector
+
+object BytesCodec {
+
+  def decode(from: BitVector, size: Int): Attempt[DecodeResult[Array[Byte]]] = {
+    val requiredBits = size * byteSize
+    if (from.length < size * byteSize) Attempt.failure(Err.insufficientBits(requiredBits, from.length))
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/codecs/binary/BytesCodec.scala:10]: you defined the variable then didn't use it

File: common/src/main/scala/co/topl/utils/codecs/binary/IntCodec.scala [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    ULongCodec.decode(from).map(result => result.map(uLongVal => decodeZigZagInt(uLongVal.toInt)))
+
+  def encode(value: Int): Attempt[BitVector] = ULongCodec.encode(encodeZigZagInt(value))
+
+  val codec: Codec[Int] = new Codec[UShort] {
+    override def decode(bits: BitVector): Attempt[DecodeResult[UShort]] = IntCodec.decode(bits)
+
+    override def encode(value: UShort): Attempt[BitVector] = IntCodec.encode(value)
+
+    override def sizeBound: SizeBound = SizeBound.atMost(64)
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/codecs/binary/IntCodec.scala:19]: 64? I thought an Int was 4 bytes (32 bits)?

File: common/src/main/scala/co/topl/utils/codecs/binary/ULongCodec.scala [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    var result: Long = 0
+    var iteration = 0
+
+    while (iteration < 10) {
+      val bitPointer = iteration * byteSize
+
+      if (from.length < bitPointer + byteSize)
+        return Attempt.failure(Err.insufficientBits(bitPointer + byteSize, from.length))
+
+      val b = from.slice(bitPointer, bitPointer + byteSize)
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/codecs/binary/ULongCodec.scala:24]: I'd really like some more comments explaining the magic in these 15 or so lines. At first glance, why is the iteration limited to 10, what is the iteration * 7 for, what is the first conditional check doing (after inspection that one is easy but a comment would make it obvious)?

File: common/src/main/scala/co/topl/utils/codecs/binary/ULongCodec.scala [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    var output = BitVector.empty
+    var runningValue = value
+
+    while (true)
+      if ((runningValue & ~0x7fL) == 0) {
+        output = output ++ BitVector(runningValue.asInstanceOf[Byte])
+        return Attempt.successful(output)
+      } else {
+        output = output ++ BitVector(((runningValue.asInstanceOf[Int] & 0x7f) | 0x80).toByte)
+        runningValue >>>= 7
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/codecs/binary/ULongCodec.scala:47]: review should also discuss this operator `>>>=`

File: common/src/main/scala/co/topl/utils/codecs/binary/SmallStringCodec.scala [hunk 9]
```diff
@@ -0,0 +1,59 @@
+package co.topl.utils.codecs.binary
+
+import cats.implicits._
+import scodec.bits.BitVector
+import scodec.{Attempt, Codec, DecodeResult, Decoder, Err, SizeBound}
+
+object SmallStringCodec {
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/codecs/binary/SmallStringCodec.scala:7]: I'm not really a fan of mixing SmallString and IntString. I agree that ByteString was a little confusing but we should either pick a different way to describe strings of varying length, or we should standardize to ByteString, ShortString, IntString, LongString, etc.

File: common/src/main/scala/co/topl/utils/codecs/binary/SmallStringCodec.scala [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+  val maxBytes: Int = 255
+
+  def decode(from: BitVector): Attempt[DecodeResult[SmallString]] =
+    Attempt.fromEither(
+      for {
+        // split input by size bits and remaining
+        sizeSplitTuple <-
+          Either.cond(from.length >= byteSize, from.splitAt(byteSize), Err.insufficientBits(byteSize, from.length))
+        stringSizeBits = sizeSplitTuple._1.toInt(signed = false) * byteSize
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/codecs/binary/SmallStringCodec.scala:17]: should a string have a positive length?

File: common/src/main/scala/co/topl/utils/IdiomaticScalaTransition.scala (hunk 1 above)
Comment (b-vennes) [reply common/src/main/scala/co/topl/utils/IdiomaticScalaTransition.scala:63]: I'll fix the toTryOps, but Option already has `.get` which is basically the same thing

File: common/src/main/scala/co/topl/utils/codecs/binary/ULongCodec.scala (hunk 7 above)
Comment (b-vennes) [reply common/src/main/scala/co/topl/utils/codecs/binary/ULongCodec.scala:24]: Unfortunately, I don't have a good explanation for the magic. I just made sure that the same couple of steps are run as before.

File: common/src/main/scala/co/topl/utils/codecs/binary/SmallStringCodec.scala (hunk 10 above)
Comment (b-vennes) [reply common/src/main/scala/co/topl/utils/codecs/binary/SmallStringCodec.scala:17]: Can you explain what you mean?

**Comment by b-vennes** [comment]
Closing branch to continue work on serialization refactor in a separate branch than `dev`. This branch will become the new `serialization-refactor` source of work.

**Pull request closed without merging**